### ISSUE TITLE
Issues with FileUpload field

### ISSUE THREAD

**Issue by iSoap** [umbraco/Umbraco.Forms.Issues#255]
Umbraco: 7.15.3
Umbraco Forms: 7.2

We just updated to the latest version 7.2, and we found the Upload Field to be very unusable as is.
We have looked into these problems and fixed a few of them by overriding the FileUpload field, but that takes away the possibility to use Perplex upload field.

Here are some of the issues with the fixes we have done.

1. ValidateField doesn't add an error even if the field has no values, and is mandatory.

It looks like you are making a check here for postedValues and not the field.Values, which is where the upload paths get added now. So you just need to add a check for that.

We have added.
if (field.Mandatory == true && field.Values.Any() == false)

That takes care of the first part. It looks like the HasCondition.IsVisible part works, but I can't be sure. We have made our own check for that as well.

2. ProcessSubmittedValue and ConvertToRecord tries to delete the field file folder in (\media\forms\upload\{formFolder}\{fieldFolder}), and fails because it can't delete a folder with values in it.

In ProcessSubmittedValue we have removed the code (line 101 to 104) because there is no reason to delete the folder.
In ConvertToRecord we have added an extra check right under the 'if (File.Exists({filePath}))' check on line 156, which checks if the current folder is a temp folder, and if it is true, then it can continue and delete the folder.

3. ProcessSubmittedValue looks at all uploaded files, and not just the one for the specific upload field.
This gives errors when having two mandatory upload fields.

Case is, if you upload to both upload fields, go a step forward, and then back again.
The upload fields show the "hidden"  uploaded files just fine.
But if you upload a new file to one of these upload fields, and try to go forward, the other one will complain that you didn't upload a file to it. 
It should be fine because if has a "hidden file", but during the processSubmittedValue it finds an uploaded file, despite it not belonging to this field, and deletes the hidden file.

For this one we have fixed the hasUploadedFile flag to check for uploaded files for this specific upload field.
It replaces line 74 to 82.

var currentFieldHasUploadedFile = files.AllKeys.Contains(field.Id.ToString()) && files[field.Id.ToString()].ContentLength > 0;

4. Has been fixed in 7.3

**Comment by chrisdent-enjoydigital** [comment]
Just to add to this, I have also found that the File Upload field type also does not work in Umbraco Grid.

We are using Umbraco v7.12.4 and Umbraco forms 7.2.0. I have added a form to a page via the Umbraco Grid data type and when the form is submitted the file upload value does not appear in the entry record.

**Comment by mov3nforward** [comment]
We have an issue where the attachment works sometimes and other times it doesn't. The form is a simple form with an attachment. The attachment is sometimes attached and others it isn't. The form is always submitted. I have received a few pdfs from people that tried to send the attachments through the from and tested myself and the files upload and attach, so it is not an issue with the file name. The form package does not work properly. We have umbraco 7.15.1 and form 7.1.1. We NEED to get this fixed as soon as possible. Our client is asking if we can revert their site back to WordPress because the form tool in their old site did not have this issue and they need this feature to work!

**Comment by nul800sebastiaan** [comment]
@iSoap / @chrisdent-enjoydigital - can you test this again in 7.3? If you're using Recaptcha you might not be able to due to this problem - https://github.com/umbraco/Umbraco.Forms.Issues/issues/252, but if you're not using Recaptcha it would be good to know if this is now fixed for you.

**Comment by iSoap** [comment]
I have removed our custom code for FileUpload, and 7.3 seems to have fixed the issues.

**Comment by chrisdent-enjoydigital** [comment]
@nul800sebastiaan After some further testing here we discovered that the issue was actually caused by our custom theme which was affecting the file upload directly. Once the theme was removed, the file upload worked fine both in and outside the grid.

For further information for you, I installed a clean version of Umbraco v7.12.4 and Umbraco forms 7.3.0 and the file upload also worked on that within the grid, which was what led me to find that the issue was not being caused by Umbraco Forms and grid.

Thanks for replying
